### ISSUE TITLE
multithreading issues with 0.9.10

### ISSUE THREAD

**Issue by schabe77** [ronmamo/reflections#81]
Hi!
I'm experiencing multithreading issues with version 0.9.10.

With two threads accessing the same jar file i get the following Log Messages:
DEBUG Thread-1 2015-07-15 11:26:33,136 org.reflections.Reflections - going to scan these urls: 
jar:file:/path/to/jar/myjar.jar!/
DEBUG Thread-2 2015-07-15 11:26:33,138 org.reflections.Reflections - going to scan these urls: ...
jar:file:/path/to/jar/myjar.jar!/
INFO Thread-1 2015-07-15 11:26:33,289 org.reflections.Reflections - Reflections took 153 ms to scan 3 urls, producing 207 keys and 529 values

Thread-2 crashes with this stack:

java.lang.IllegalStateException: zip file closed
        at java.util.zip.ZipFile.ensureOpen(ZipFile.java:634)
        at java.util.zip.ZipFile.access$200(ZipFile.java:56)
        at java.util.zip.ZipFile$1.hasMoreElements(ZipFile.java:487)
        at java.util.jar.JarFile$1.hasMoreElements(JarFile.java:241)
        at org.reflections.vfs.ZipDir$1$1.computeNext(ZipDir.java:30)
        at org.reflections.vfs.ZipDir$1$1.computeNext(ZipDir.java:26)
        at com.google.common.collect.AbstractIterator.tryToComputeNext(AbstractIterator.java:143)
        at com.google.common.collect.AbstractIterator.hasNext(AbstractIterator.java:138)
        at org.reflections.Reflections.scan(Reflections.java:240)
        at org.reflections.Reflections.scan(Reflections.java:204)
        at org.reflections.Reflections.<init>(Reflections.java:129)
        at org.reflections.Reflections.<init>(Reflections.java:170)
        at org.reflections.Reflections.<init>(Reflections.java:143) 

The ZipDir seems to be created in DefaultUrlTypes.jarUrl.createDir(URL):
    return new ZipDir(((JarURLConnection) urlConnection).getJarFile()); 

but sun.net.www.protocol.jar.JarURLConnection.getJarFile() uses (or may use) a cache:
  this.jarFile = factory.get(getJarFileURL(), getUseCaches()); 

In my case both threads get the same instance of JarFile, and after the first thread has closed the ZipDir (which closes the JarFile) in 
  org.reflections.Reflections.scan(URL):260
the second one gets the IllegalStateException.

Cheers,

Christian

**Comment by schabe77** [comment]
A fix could be to disable JarURLConnection's caching in DefaultUrlTypes.jarUrl.createDir(URL):
            if (connection instanceof JarURLConnection) {
                JarURLConnection jarURLConnection =  (JarURLConnection) connection;
                jarURLConnection.setUseCaches(false);
                return new ZipDir(jarURLConnection.getJarFile());
            }

**Comment by schabe77** [comment]
A fix could be to disable URLConnection's caching in DefaultUrlTypes.jarUrl.createDir(URL):

            URLConnection urlConnection = url.openConnection();
            urlConnection.setUseCaches(false);
            if (urlConnection instanceof JarURLConnection) {
                    return new ZipDir(((JarURLConnection) urlConnection).getJarFile());
            }

**Comment by cteyton** [comment]
I got a similar issue !

**Comment by MrMorden** [comment]
I can see this in our development environment (Windows) but not in our target environment (Linux).
Quite easy to reproduce using two threads.
Problem does not exist in 0.9.9.

**Comment by jenspopp** [comment]
I can reproduce the issue with reflections in a wildfly environment.

**Comment by collinscangarella** [comment]
Reproduced here as well, both on OSX and RHEL. For some reason RHEL doesn't throw the exception - it just hangs or silently fails (not sure which). Our workaround is to use v0.9.9.

**Comment by corrspt** [comment]
I think I got hit with the same problem. While using a Play Framework application that uses the Katharsis Library (which in turn uses reflections). I downgraded to 0.9.9 (some other dependency updated to 0.9.10 and the problem started)

**Comment by lukehutch** [comment]
As pointed out in the first comment, this is indicative of a single java.util.zip.ZipFile instance being used in multiple threads.

This is not only a problem when one thread closes the ZipFile, which gave rise to this bug report, but also ZipFile actually has an internal lock that only grants access to one thread at a time (so Reflections is not making optimal use of CPU resources).

**Comment by jimfcarroll** [comment]
I got the same problem with 0.9.11

**Comment by LuboVarga** [comment]
Same for us. Problem is presented in 0.9.11 also. Version 0.9.9 is working OK. We use it in maven mojo and when running parallel build, exception happens.

**Comment by lukehutch** [comment]
@LuboVarga My apologies to the author of Reflections for the promotion of my own project, but since this bug has been open for almost 2 years now... if this bug is a show-stopper for you, you could try [FastClasspathScanner](https://github.com/lukehutch/fast-classpath-scanner/) instead. It opens one ZipFile instance per zipfile per thread, so it does not suffer from this problem.

**Comment by wangzhenxyz1** [comment]
I encounter the same problem version 0.9.11 on the window platform in multi-threads env.

**Comment by andreabertagnolli** [comment]
the regression was introduced with this PR: https://github.com/ronmamo/reflections/pull/57
i'm working on a patch

**Comment by lukehutch** [comment]
@andreabertagnolli the solution is to create a separate `ZipFile` instance for each thread -- this is what I had to do in FastClasspathScanner, which also does multithreaded scanning.

The exception on close issue is not the only problem here: without creating one `ZipFile` instance per thread, multithreading doesn't even speed up the scanning process, because all the methods within `ZipFile` are locked on a per-zipfile basis, so only one thread can access a given jar at one time.

**Comment by andreabertagnolli** [comment]
You're right, but my PR is just a fix, it simply avoids the exception.
I never thought about performance, it's not the scope of this issue ;)

Il mar 7 nov 2017, 04:17 Luke Hutchison <notifications@github.com> ha
scritto:

**Comment by ronmamo** [comment]
for the context, accessing the same jar concurrently from multiple threads sounds like something wrong is done. the better approach is to have one reflections instance scanning only once, with all relevant scanners and configuration. parallel scanning means each url scan goes in different thread, so there should not be contention. 
this pr merged however, as it solves this misuse - correct me if I'm wrong.

@lukehutch promoting your open source is totally acceptable! sharing goes both ways, this is open source : )

**Comment by lukehutch** [comment]
@ronmamo Thanks for the pragmatic attitude :-)

@ronmamo @andreabertagnolli the PR should fix the issue, but at the cost of unzipping the same jars-within-jars multiple times, once per thread.

With the situation before this PR, the first thread to finish scanning would close the `ZipFile`, then other threads would not be able to complete their scanning of the jar, because the `ZipFile` instance was shared. The jar-within-jar was only extracted once, but only one thread could actually read from the jar at a time anyway (due to locking). With the new PR, parallel scanning is possible, but work is duplicated.

**Comment by ronmamo** [comment]
@lukehutch why would you scan the same jar twice and/or concurrently, again?

**Comment by lukehutch** [comment]
@ronmamo I was under the impression that you were parallelizing the scanning work within a given jarfile (in this case, for extracting jars from within jars, e.g. in the JBoss VFS, and assigning a different scanning thread to each nested jar (or simply assigning different threads to scan each classfile within a jar). Isn't this what was triggering the exception, when the first thread to finish scanning files in a jar closed the jar, triggering an exception in the other threads?

**Comment by karianna** [comment]
@lukehutch Thanks for the fix, we'd like to use the patched version - any chance of a 0.9.12 release soon?

**Comment by lukehutch** [comment]
@karianna I'm not the developer of Reflections (I created FastClasspathScanner). @ronmamo created Reflections, so he would need to answer that question.

**Comment by yesudev** [comment]
@ronmamo Please let us know when the patch will be released with the fix. We are stopping due to this issue.

**Comment by lukehutch** [comment]
I posted an explanation to a related bug about how to access `ZipFile` in a multithreaded environment. I'll repost here, since a lot of things link to this bug.

--

If you want multiple threads to read from the same zipfile in a scalable way, you must open one `ZipFile` instance per thread. That way, the per-thread lock in the `ZipFile` methods does not block all but one thread from reading from the zipfile at one time. It also means that when each thread closes the `ZipFile` after they're done reading, they close their own instance, not the shared instance, so you don't get an exception on the second and subsequent close.

Protip: if you really care about speed, you can get more performance by reading all the `ZipEntry` objects from the first ZipFile instance, and sharing them with all threads, to avoid duplicating work in reading `ZipEntry` objects for each thread separately. A `ZipEntry` object is not tied to a specific `ZipFile` instance per se, `ZipEntry` just records metadata that will work with any `ZipFile` object representing the same zipfile that the `ZipEntry` came from.

**Comment by cedricschweizer** [comment]
I had the same problem with version 0.9.11. After updating to 0.9.9 everything works fine for me!

**Comment by jaeyeonling** [comment]
We are having this problem in 0.9.11

**Comment by 455586841** [comment]
<dependency>
            <groupId>org.reflections</groupId>
            <artifactId>reflections</artifactId>
            <version>0.9.12</version>
        </dependency>


 Invocation of init method failed; nested exception is java.lang.IllegalStateException: zip file closed
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1762)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:593)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:515)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:320)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:318)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:849)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:877)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:549)
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:142)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:775)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:397)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:316)
	at com.ruibo.sys.web.Application.main(Application.java:37)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:50)
	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:51)
Caused by: java.lang.IllegalStateException: zip file closed
	at java.util.zip.ZipFile.ensureOpen(ZipFile.java:686)
	at java.util.zip.ZipFile.access$200(ZipFile.java:60)
	at java.util.zip.ZipFile$ZipEntryIterator.<init>(ZipFile.java:499)
	at java.util.zip.ZipFile.entries(ZipFile.java:552)
	at java.util.jar.JarFile.access$001(JarFile.java:71)
	at java.util.jar.JarFile$JarEntryIterator.<init>(JarFile.java:250)
	at java.util.jar.JarFile$JarEntryIterator.<init>(JarFile.java:247)
	at java.util.jar.JarFile.stream(JarFile.java:279)
	at org.reflections.vfs.ZipDir.lambda$getFiles$2(ZipDir.java:21)
	at org.reflections.Reflections.scan(Reflections.java:260)
	at org.reflections.Reflections.scan(Reflections.java:216)
	at org.reflections.Reflections.<init>(Reflections.java:137)
	at com.ruibo.canal.spring.event.AllTablePublicColumnSyncUpdateApplicationListener.afterPropertiesSet(AllTablePublicColumnSyncUpdateApplicationListener.java:113)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1821)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1758)

**Comment by Numbernick** [comment]
I'm also having this issue with 0.9.12. Switching back to 0.9.9 solved the problem.

**Comment by KlemenDEV** [comment]
Are there any plans to release this fix in a new release?

**Comment by chenchaoyun0** [comment]
> I'm also having this issue with 0.9.12. Switching back to 0.9.9 solved the problem.

thx
